src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,399 @@
+
+	m_sending = true;
+
+	tracker_message msg;
+    std::weak_ptr<request_callback> cb;
+    std::tie(msg, cb) = m_pending.front();
+
+	// Update requester
+	if(cb.lock()) m_requester = cb;
+
+	std::visit([&](auto const& m)
+		{
+			do_send(m);
+		}
+		, msg
+	);
+}
+
+void websocket_tracker_connection::do_send(tracker_request const& req)
+{
+	json payload;
+	payload["action"] = "announce";
+	payload["info_hash"] = from_latin1({req.info_hash.data(), std::size_t(req.info_hash.size())});
+	payload["uploaded"] = req.uploaded;
+	payload["downloaded"] = req.downloaded;
+	payload["left"] = req.left;
+	payload["corrupt"] = req.corrupt;
+	payload["numwant"] = req.num_want;
+
+	char str_key[9];
+	std::snprintf(str_key, sizeof(str_key), "%08X", req.key);
+	payload["key"] = str_key;
+
+	static const char* event_string[] = {"completed", "started", "stopped", "paused"};
+	if(req.event != event_t::none)
+		payload["event"] = event_string[static_cast<int>(req.event) - 1];
+
+	payload["peer_id"] = from_latin1({req.pid.data(), req.pid.size()});
+
+	payload["offers"] = json::array();
+	for(auto const& offer : req.offers)
+	{
+		json payload_offer;
+		payload_offer["offer_id"] = from_latin1({offer.id.data(), offer.id.size()});
+		payload_offer["offer"]["type"] = "offer";
+		payload_offer["offer"]["sdp"] = offer.sdp;
+		payload["offers"].push_back(payload_offer);
+	}
+
+	std::string const data = payload.dump();
+
+	std::shared_ptr<request_callback> cb = requester();
+	if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_WRITE [ %s ]", data.c_str());
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+    m_websocket->async_write_some(boost::asio::const_buffer(data.data(), data.size())
+     		, std::bind(&websocket_tracker_connection::on_write, self, _1, _2));
+}
+
+void websocket_tracker_connection::do_send(tracker_answer const& ans)
+{
+    json payload;
+    payload["action"] = "announce";
+    payload["info_hash"] = from_latin1({ans.info_hash.data(), std::size_t(ans.info_hash.size())});
+    payload["offer_id"] = from_latin1({ans.answer.offer_id.data(), ans.answer.offer_id.size()});
+    payload["to_peer_id"] = from_latin1({ans.answer.pid.data(), ans.answer.pid.size()});
+    payload["peer_id"] =  from_latin1({ans.pid.data(), ans.pid.size()});
+    payload["answer"]["type"] = "answer";
+    payload["answer"]["sdp"] = ans.answer.sdp;
+
+	std::string const data = payload.dump();
+
+	std::shared_ptr<request_callback> cb = requester();
+	if (cb) cb->debug_log("*** WEBSOCKET_TRACKER_WRITE [ %s ]", data.c_str());
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+	m_websocket->async_write_some(boost::asio::const_buffer(data.data(), data.size())
+			, std::bind(&websocket_tracker_connection::on_write, self, _1, _2));
+}
+
+void websocket_tracker_connection::do_read()
+{
+	m_read_buffer.clear();
+
+	std::shared_ptr<websocket_tracker_connection> self(shared_from_this());
+	m_websocket->async_read_some(m_read_buffer
```

Comment:
is `async_read_some()` really correct here? it looks like `on_read()` expects a full message. Shouldn't it be `async_read()`?